packages/merkle-trees/src/sparse_merkle.nr

```diff
@@ -1,112 +1,208 @@
-        leaf_hasher: fn([Field; 3]) -> Field,
-        hasher: fn([Field; 2]) -> Field,
-    ) -> Self {
+impl<T> SMT_Creator<T> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
+    fn default(root: T, leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
         Self { root, leaf_hasher, hasher }
     }
-}
 
-impl MembershipProver<(Field, Field), Field, Field> for SparseMerkleTree {
-    fn membership<let N: u32>(self, entry: (Field, Field), indexes: Field, hash_path: [Field; N]) {
-        let mut calculated_root: Field = 0;
+    fn from(root: T, leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
+        Self::default(root, leaf_hasher, hasher)
+    }
+
+    fn new(leaf_hasher: fn([T; 3]) -> T, hasher: fn([T; 2]) -> T) -> Self {
+        Self::from(T::default(), leaf_hasher, hasher)
+    }
+}
 
+impl<T> MembershipProver<T, (T, T)> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
+    fn membership<let N: u32>(self, entry: (T, T), indexes: Field, hash_path: [T; N]) {
         // membership proof: the root is calculated based on the entry, the siblings,
         // and the path determined by the key of entry through consecutive hashing
-        calculated_root = self.calculate_root(entry, indexes, hash_path);
-
-        assert(calculated_root == self.root);
+        assert(self.calculate_root(entry, indexes, hash_path) == self.root);
     }
 }
 
-impl NonMembershipProver for SparseMerkleTree {
+impl<T> NonMembershipProver<T, (T, T)> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
     fn non_membership(
         self,
-        entry: (Field, Field),
-        matching_entry: (Field, Field),
-        siblings: [Field; 256],
+        entry: (T, T),
+        matching_entry: (T, T),
+        index: Field,
+        siblings: [T; 256],
     ) {
-        if (self.root != 0) {
+        if (self.root != T::default()) {
             // non-membership proof: the root is calculated based on the matching_entry, the siblings
             // and the path that is determined by the key of entry. This makes sure that matching_entry is in fact
             // a matching entry for entry meaning that it shares the same first bits as path
-            let mut calculated_root = self.calculate_root(matching_entry, entry.0, siblings);
+            let mut calculated_root = self.calculate_root(matching_entry, index, siblings);
             assert(calculated_root == self.root);
         }
     }
 }
 
-impl Modifier<(Field, Field), Field, Field> for SparseMerkleTree {
-    fn add<let N: u32>(
-        &mut self,
-        new_entry: (Field, Field),
-        indexes: Field,
-        hash_path: [Field; N],
-    ) {
+impl<T> Modifier<T, (T, T)> for SparseMerkleTree<T>
+where
+    T: Eq + Default,
+{
+    fn add<let N: u32>(&mut self, new_entry: (T, T), index: Field, hash_path: [T; N]) {
         // if the root node is zero the first leaf is added to the tree in which case
         // the new root equals H(k,v,1)
```

Comment:
Yes initially I wrote a Trait that ensured T implemented `T::one()`. Then I did some research and found out that although 1 was chosen as leaf marker on the zk-kit implementation, it actually just checked if the third position of the child node was "truthy". So I made [the tiniest PR](https://github.com/privacy-scaling-explorations/zk-kit/pull/354/files) to specifically assert for `undefined`, which allows leaves to be marked with whatever value is truthy. This is very useful since we don'can't know all the possible implementations for `T::default()`.

I'll add a clarification on the README, although I don't think it's too relevant for Noir as long as your type implements Default (the LSP will scream at you if you don't). It is relevant if you're also using the zk-kit SMT lib, though